Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,501 @@
+                                                                attachment.error = .couldNotConvertToJpeg
+                                                                return attachment
+            }
+
+            if jpgImageData.count <= kMaxFileSizeImage {
+                let recompressedAttachment = SignalAttachment(data : jpgImageData, dataUTI: kUTTypeJPEG as String)
+                recompressedAttachment.image = dstImage
+                return recompressedAttachment
+            }
+
+            // If the JPEG output is larger than the file size limit,
+            // continue to try again by progressively reducing the
+            // image upload quality.
+            switch imageUploadQuality {
+            case .uncropped:
+                imageUploadQuality = .high
+            case .high:
+                imageUploadQuality = .medium
+            case .medium:
+                imageUploadQuality = .low
+            case .low:
+                attachment.error = .fileSizeTooLarge
+                return attachment
+            }
+        }
+    }
+
+    private class func imageScaled(_ image: UIImage, toMaxSize size: CGFloat) -> UIImage {
+        var scaleFactor: CGFloat
+        let aspectRatio: CGFloat = image.size.height / image.size.width
+        if aspectRatio > 1 {
+            scaleFactor = size / image.size.width
+        } else {
+            scaleFactor = size / image.size.height
+        }
+        let newSize = CGSize(width: CGFloat(image.size.width * scaleFactor), height: CGFloat(image.size.height * scaleFactor))
+        UIGraphicsBeginImageContext(newSize)
+        image.draw(in: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(newSize.width), height: CGFloat(newSize.height)))
+        let updatedImage: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        return updatedImage!
+    }
+
+    private class func maxSizeForImage(image: UIImage, imageUploadQuality: TSImageQuality) -> CGFloat {
+        switch imageUploadQuality {
+        case .uncropped:
+            return max(image.size.width, image.size.height)
+        case .high:
+            return 2048
+        case .medium:
+            return 1024
+        case .low:
+            return 512
+        }
+    }
+
+    private class func jpegCompressionQuality(imageUploadQuality: TSImageQuality) -> CGFloat {
+        switch imageUploadQuality {
+        case .uncropped:
+            return 1
+        case .high:
+            return 0.9
+        case .medium:
+            return 0.5
+        case .low:
+            return 0.3
+        }
+    }
+
+    // MARK: Video Attachments
+
+    // Factory method for video attachments.
+    //
+    // NOTE: The attachment returned by this method may not be valid.
+    //       Check the attachment's error property.
+    public class func videoAttachment(data: Data?, dataUTI: String) -> SignalAttachment {
```

Comment:
nit: looks like we're already checking if-nil at the caller, so we can make these `data` parameters non-optional.